[myenergi] Initial contribution

renescherer opened the pull request:
# myenergi Binding

This binding supports a number of devices provided by the company [myenergi](https://www.myenergi.com/). They include the Zappi car charger (EVSE), the Eddi solar diverter and the Harvi remote sensor for CT clamps.

    [myenergi] Initial contribution to support Zappi, Eddi and Harvi devices

## Description

Although this is the first official PR for this binding, it has been in the market place for a while for 3.x and 4.x openHAB systems and has been used, tested and extended by various contributors. This is the latest version which aligns with the standard naming conventions and works with openHAB 5.x.

Details about the binding can be found in the [market place](https://community.openhab.org/t/myenergi-myenergi-binding/139667)

## Testing

Your pull request will automatically be built and available under the following folder:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

renescherer committed:
[myenergi] initial release of myenergi binding for openHAB 5.x

Signed-off-by: Rene Scherer <rene@scherer-online.com>

renescherer committed:
[myenergi] resolved a couple of Copilot issues (typos/copy-paste errors)

Signed-off-by: Rene Scherer <rene@scherer-online.com>

lsiepel commented on bundles/org.openhab.binding.myenergi/src/test/resources/jetty-logging.properties:
What is the purpose of this file? I have not seen other bindings using this.

renescherer replied:
I used this when I did some debugging around digest authentication with Jetty client. I've removed it now.

lsiepel commented on bundles/org.openhab.binding.myenergi/src/test/resources/log4j.properties:
What is the purpose of this file? I have not seen other bindings using this.

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,453 @@
+
+	<channel-type id="clamp-name-type">
+		<item-type>String</item-type>
+		<label>Clamp Name/Type</label>
```
Name or Type, pick one :-)

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,453 @@
+		<state readOnly="true" pattern="%d %%"/>
+	</channel-type>
+
+	<channel-type id="boost-end-time-type">
```
Is this a timestamp or a remaining duration? Either wya, better to make use of UoM

lsiepel reviewed (changes requested):
Thanks for contributing this binding. I left quite some comments, don't let it demotivate you, in general it looks good, it just are many small things to look at. I do have one slightly bigger comment burried somewhere about the seperation of concern regarding the bridge and its responsibilities. Let me know if any of my comments are unclear, it took me some time to review ;-)
When these are addressed, i can look at the remaining files (handlers + apiclient)

renescherer commented on bundles/org.openhab.binding.myenergi/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,453 @@
+		<description>myenergi wireless energy sensor</description>
+
+		<channels>
+			<channel id="last-updated-time" typeId="last-updated-time-type"/>
+			<channel id="clamp-name1" typeId="clamp-name-type"/>
+			<channel id="clamp-power1" typeId="clamp-power-type"/>
+			<channel id="clamp-phase1" typeId="phase-type"/>
+			<channel id="clamp-name2" typeId="clamp-name-type"/>
+			<channel id="clamp-power2" typeId="clamp-power-type"/>
+			<channel id="clamp-phase2" typeId="phase-type"/>
+			<channel id="clamp-name3" typeId="clamp-name-type"/>
+			<channel id="clamp-power3" typeId="clamp-power-type"/>
+			<channel id="clamp-phase3" typeId="phase-type"/>
```
I know they don't need the -type suffix, but I think it helps when putting the xml file together. To me it was more readable when you can easily distinguish between the channel name and the channel type name. I've remove the suffix now.

renescherer commented on bundles/org.openhab.binding.myenergi/README.md:
```diff
@@ -0,0 +1,225 @@
+| plug-status         | String                   | Describes state of connection to EV                             |
+| command-tries       | Number                   | 0-10 Trying, 253 Failed, 254 OK, 255 No command sent            |
+| diverter-priority   | Number                   | Priority for solar energy relative to other myenergy devices    |
+| minimum-green-level | Number                   | Minimum proportion of solar energy used                         |
```
updated the comment and made the type Number:Dimensionless

renescherer commented on bundles/org.openhab.binding.myenergi/README.md:
````diff
@@ -0,0 +1,225 @@
+
+myenergi.things
+
+```
````
why java?

lsiepel replied:
see other comment. For code highlighting

renescherer commented:
Thanks very much @lsiepel for the detailed review. I've started work on the changes and hopefully will have a new version available sometime next week.

renescherer commented on bundles/org.openhab.binding.myenergi/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,453 @@
+		<label>Locking Mode</label>
+		<description>The current locking mode of the device as bitmap.Bit 0: Locked Now, Bit 1: Lock when plugged in, Bit 2:
+			Lock when unplugged, Bit 3: Charge when locked, Bit 4: Charge Session Allowed (Even if locked)</description>
+		<state readOnly="true" pattern="%d"/>
```
The value is returned as a bitmap. I.e. multiple bits can be set concurrently. Theoretically I could present it as 5 individual switches. Or is there another way to show a multi-switch context?

lsiepel replied:
Looking at the values, i have the impression that some are mutually exclusive and some are not.
This should be properly modelled to openHAB. Maybe multiple channels. 

Lock when plugged in / Lock when unplugged seems mutally exclusive or not?
All others can be modelled as a switch as they can be set independent.

renescherer commented on bundles/org.openhab.binding.myenergi/README.md:
```diff
@@ -0,0 +1,225 @@
+| charging-mode       | Number                   | Boost (0), Fast (1), Eco (2), Eco+ (3) or Stop (4)              |
+| status              | Number                   | Describes what the charger is doing                             |
+| plug-status         | String                   | Describes state of connection to EV                             |
+| command-tries       | Number                   | 0-10 Trying, 253 Failed, 254 OK, 255 No command sent            |
```
An API call can succeed and update the value at the myenergi server, it doesn't necessarily mean that the command has been successfully transferred to the charger. This field will show you minutes later if it's still retrying or succeeded. I certainly would like to see this on a UI, not just in the openhab logs.

renescherer replied:
yes, only the most recent. I could rename it to something like "last-command-retry-status" to make it more obvious. The current name was based on what Myenergi calls it in the command result.
I do think it's useful for the user to understand if the latest command has somehow reached the charger.

renescherer replied:
change into a string status message

renescherer commented on bundles/org.openhab.binding.myenergi/README.md:
```diff
@@ -0,0 +1,225 @@
+| grid-power          | Number:Power             | Power taken from grid                                           |
+| generated-power     | Number:Power             | Power generated by solar panels                                 |
+| diverted-power      | Number:Power             | power used by charger                                           |
+| charge-added        | Number:Energy            | charge added in this or the last charing session                |
```
changed description to: amount of energy added in this or the last charging session

renescherer commented on bundles/org.openhab.binding.myenergi/README.md:
```diff
@@ -0,0 +1,225 @@
+| last-updated-time   | DateTime                 | The time the readings have last been updated.                   |
+| supply-voltage      | Number:ElectricPotential | The supply voltage to the Zappi                                 |
+| supply-frequency    | Number:Energy            | The amount of energy measured by the clamp.                     |
+| number-of-phases    | Number                   | The identifier of the phase (for 3-phase installations).        |
```
The number of phases for this installation (either 1 or 3). I've moved it to properties.

renescherer commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/MyenergiHandlerFactory.java:
```diff
@@ -0,0 +1,89 @@
+        return null;
+    }
+
+    @Reference
+    protected void setHttpClientFactory(HttpClientFactory httpClientFactory) {
+        apiClient.setHttpClientFactory(httpClientFactory);
+    }
+
+    protected void unsetHttpClientFactory(HttpClientFactory httpClientFactory) {
+        apiClient.setHttpClientFactory(null);
+    }
```
Not sure I understand your second comment. At the moment, I'm not using getCommonHttpClient() anywhere apart from the Test classes. I'm creating a customer client using createHttpClient in MyenergiApiClient:
```
// create a new httpClient, so that we can add our own digest authentication
client = factory.createHttpClient(MyenergiApiClient.class.getSimpleName());
AuthenticationStore auth = client.getAuthenticationStore();

```
This is only done once during the initialization of the API client.

lsiepel replied:
1. Remove these two methods and use the constructor:  
    ```java
    @Activate
    public MyenergiHandlerFactory(final @Reference HttpClientFactory httpClientFactory) {
        this.httpClientFactory= httpClientFactory;
    }
    ```
2. Change the APIClient to not hold the httpfactory. You can change the constructor and set the httpclient from the factory there.
```java
    public MyenergiApiClient(final HttpClientFactory httpClientFactory) {
        client = httpClientFactory.createHttpClient(MyenergiApiClient.class.getSimpleName());
    }
```

Also relates to https://github.com/openhab/openhab-addons/pull/19248/files#r2401467701

renescherer commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/dto/BaseSummary.java:
```diff
@@ -0,0 +1,95 @@
+    public ZonedDateTime getLastUpdateTime() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm:ss");
+        LocalDateTime ldt = LocalDateTime.parse(dat + " " + tim, formatter);
+        ZonedDateTime dateTime = ZonedDateTime.of(ldt, ZoneId.systemDefault());
```
The last updated time is set by the binding, not the remote service. The API doesn't actually return any absolute times anywhere. That's why I went for the system default.

lsiepel replied:
This can be simplified by using an `Instant` if needed the `TimeZoneProvider` should be used so it respects the set timezoen by the openHAB configuration.

renescherer replied:
actually, reading the docs again, this is always returned in UTC. Change code accordingly.

renescherer commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/dto/MyenergiData.java:
```diff
@@ -0,0 +1,165 @@
+        return harvis;
+    }
+
+    public synchronized void setHarvis(List<HarviSummary> harvis) {
```
That was unnecessary. I've removed the synchronization.

lsiepel replied:
It is still there, and also on many other dto objects.

renescherer committed:
various changes from recent code review

Signed-off-by: Rene Scherer <rene@scherer-online.com>

renescherer committed:
reinstated accidentially deleted .gitignore file

Signed-off-by: Rene Scherer <rene@scherer-online.com>

renescherer committed:
fixed issue with http client initializing/starting in certain circumstances.

Signed-off-by: Rene Scherer <rene@scherer-online.com>

renescherer committed:
fixed examples of item configurations in README.md

Signed-off-by: Rene Scherer <rene@scherer-online.com>

renescherer commented:
@lsiepel I've addressed most of your suggestions above. There are few where I had some further questions and I haven't restructured the files yet, so that it's easier to trace through the initial comments.

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/dto/ZappiSummary.java:
```diff
@@ -0,0 +1,139 @@
+        super(serialNumber);
+    }
+
+    @Override
+    public Map<@NonNull String, String> getThingProperties() {
+        Map<@NonNull String, String> properties = super.getThingProperties();
+        properties.put(MyenergiBindingConstants.PROP_NUMBER_OF_PHASES, String.valueOf(numberOfPhases));
+        return properties;
+    }
```
See other comment about extracting this to a mapper

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/dto/DaysOfWeekMap.java:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+
+    public synchronized void setDay(DayOfWeek activeDay, boolean active) {
```
I wonder why this is synchronized.

lsiepel commented:
Updated open comments and added a few.

renescherer committed:
various code review changes, locking mode channel split into 4 separate channels

Signed-off-by: Rene Scherer <rene@scherer-online.com>

renescherer committed:
removed non-existent channel from readme

Signed-off-by: Rene Scherer <rene@scherer-online.com>

renescherer committed:
fixed minor issue with too long HttpClient name

Signed-off-by: Rene Scherer <rene@scherer-online.com>

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/MyenergiApiClient.java:
```diff
@@ -0,0 +1,494 @@
+
+                            logger.debug("sending API request attempt# {}: {}", innerLoop, url.toString());
+
+                            ContentResponse response = request.send();
```
Add a explicit timeout to the request

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/MyenergiDiscoveryService.java:
```diff
@@ -0,0 +1,176 @@
+        ThingUID thingsUID = new ThingUID(THING_TYPE_EDDI, bridgeUID, serialNumber.toString());
+        Map<String, Object> properties = new HashMap<String, Object>();
+        properties.putAll(ThingPropertyMapper.getThingProperties(device));
+        thingDiscovered(DiscoveryResultBuilder.create(thingsUID).withLabel("MyEnergi Eddi").withProperties(properties)
```
Add representation property

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/handler/MyenergiBaseDeviceHandler.java:
```diff
@@ -0,0 +1,300 @@
+        serialNumber = Long.parseLong(getThing().getUID().getId());
+
+        MyenergiDeviceConfiguration config = getConfigAs(MyenergiDeviceConfiguration.class);
+        if (config.refreshInterval < 10) {
```
thing-type.xml allows 5 and greather, please align, whatever value is best.
Also better to extract this to a constant.

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/handler/MyenergiBaseDeviceHandler.java:
```diff
@@ -0,0 +1,300 @@
+    public synchronized void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        logger.info("bridgeStatusChanged for {}. Reseting handler", this.getThing().getUID());
+        logger.info("bridgeStatusChanged to: {}", bridgeStatusInfo.getStatus().toString());
+        logger.info("bridgeStatusChanged to: {}", bridgeStatusInfo.getStatus().toString());
```
Remove this logging. If logging at all it should be a single debug line.

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/handler/MyenergiBridgeHandler.java:
```diff
@@ -0,0 +1,195 @@
+        logger.debug("Initializing MyEnergiBridgeHandler");
+        MyenergiBridgeConfiguration config = getConfigAs(MyenergiBridgeConfiguration.class);
+
+        if (config.hubSerialNumber.isEmpty() || config.password.isEmpty()) {
```
```suggestion
        if (config.hubSerialNumber.isBlank() || config.password.isBlank()) {
```

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/handler/MyenergiBridgeHandler.java:
```diff
@@ -0,0 +1,195 @@
+        updateStatus(ThingStatus.UNKNOWN);
+
+        try {
+            logger.debug("Login to MyEnergi API with hubSerialNumber: {}", config.hubSerialNumber);
```
This type of logging should be part of the apiclient.  (seperation of concern) 

This class does not care if it needs to login, jump trough a ring of fire or anything else, it just calls the class to initialize.

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/handler/MyenergiBridgeHandler.java:
```diff
@@ -0,0 +1,195 @@
+            logger.debug("Login to MyEnergi API with hubSerialNumber: {}", config.hubSerialNumber);
+            apiClient.initialize(config.hubSerialNumber, config.password);
+            apiClient.updateTopologyCache();
+            logger.debug("Cache update successful, setting bridge status to ONLINE");
```
Duplicate of next line, don;t do this kind of logging.

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/handler/MyenergiBridgeHandler.java:
```diff
@@ -0,0 +1,195 @@
+            apiClient.updateTopologyCache();
+            logger.debug("Cache update successful, setting bridge status to ONLINE");
+            updateStatus(ThingStatus.ONLINE);
+        } catch (AuthenticationException e) {
```
This triggers me. This method should be non-clocking.
How could an async thread produce a authenticationexception here? Am i wrong or is there a blocking call?

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/handler/MyenergiEddiActions.java:
```diff
@@ -0,0 +1,106 @@
+
+    public static void setBoost(@Nullable ThingActions actions, int heater, int duration) {
+        if (actions instanceof MyenergiEddiActions) {
+            ((MyenergiEddiActions) actions).setBoost(heater, duration);
```
```suggestion
        if (actions instanceof MyenergiEddiActions actionHandler) {
            actionHandler.setBoost(heater, duration);
```

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/handler/MyenergiEddiActions.java:
```diff
@@ -0,0 +1,106 @@
+
+    public static void cancelBoost(@Nullable ThingActions actions, int heater) {
+        if (actions instanceof MyenergiEddiActions) {
+            ((MyenergiEddiActions) actions).cancelBoost(heater);
```
```suggestion
        if (actions instanceof MyenergiEddiActions actionHandler) {
            actionHandler.cancelBoost(heater);
```

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/handler/MyenergiZappiHandler.java:
```diff
@@ -0,0 +1,275 @@
+import static org.openhab.binding.myenergi.internal.MyenergiBindingConstants.ZAPPI_CHANNEL_SUPPLY_FREQUENCY;
+import static org.openhab.binding.myenergi.internal.MyenergiBindingConstants.ZAPPI_CHANNEL_SUPPLY_VOLTAGE;
+import static org.openhab.binding.myenergi.internal.MyenergiBindingConstants.ZAPPI_CHANNEL_TIMED_BOOST_CHARGE;
+import static org.openhab.binding.myenergi.internal.MyenergiBindingConstants.ZAPPI_CHANNEL_TIMED_BOOST_DURATION;
```
Wildcard imports are allowed for static imports like this:
```suggestion
import static org.openhab.binding.myenergi.internal.MyenergiBindingConstants.*;
```

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/java/org/openhab/binding/myenergi/internal/handler/MyenergiZappiHandler.java:
```diff
@@ -0,0 +1,275 @@
+                        bh.setZappiChargingMode(serialNumber, newMode);
+                    } catch (InvalidDataException e) {
+                        logger.info("Invalid ZAPPI_CHANNEL_CHARGING_MODE update received, mode was: {} ({})",
+                                command.toString(), command.getClass().toString());
```
warn or debug. info is unexpected here.

lsiepel commented on bundles/org.openhab.binding.myenergi/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,455 @@
+		<item-type>DateTime</item-type>
+		<label>Last Updated Time</label>
+		<description>The time when the device was last updated</description>
+		<state readOnly="true" pattern="%1$tF %1$tR"/>
```
Can you add semantic tags (point and optional property) to all channel type's?

For reference: 
https://www.openhab.org/docs/developer/bindings/thing-xml.html#default-point-and-property-tags

lsiepel reviewed (changes requested):
Thanks. I have now covered all files and think this is the final review. Please also check:

- no build warnings
- no SAT warnings
- updated i18n file after all adjustments

lsiepel commented:
@renescherer JFTR if you want it in 5.1.0, we have about 4 weeks left.

lsiepel commented:
Gentle ping @renescherer can you proceed with this? You need anything, except for time :-/

lsiepel commented:
We are very close (10 days) before the feature/code freeze period starts. This PR is allmost at the finishline, would be nice to have it added in 5.2.0 @renescherer
